### ISSUE TITLE
About hybrid loss

### ISSUE THREAD

**Issue by JawadTawhidi** [hamidriasat/UNet-3-Plus#7]
Hi, sorry for disturbing you. Do you have any pytorch code for hybrid loss you have used here? if so, I would be really grateful if you share it with me.

**Comment by hamidriasat** [comment]
Hi,
[Here](https://github.com/ZJUGiveLab/UNet-Version/tree/master/loss) you can find the implementation of UNet3+ hybrid loss functions in PyTorch.

**Comment by JawadTawhidi** [comment]
Hi, I did not find the focal loss in that list.

**Comment by hamidriasat** [comment]
Check [here](https://github.com/Owais-Ansari/Unet3plus/blob/main/utils/losses.py).

**Comment by JawadTawhidi** [comment]
sorry for disturbing again, in the paper they say their hybrid loss is for 3 level hierarchy – pixel,  patch and map-level, which is able to capture both large-scale and fine structures with clear boundaries.

As I am new in the area of image segmentation, I am not familiar with these loss functions. Would you please just tell me among those 3 losses which one is for pixel level, which one is for patch level and which one is for map level? which of these loses is specifically for capturing large scale and which one is for capturing fine structure with clear boundaries?

**Comment by hamidriasat** [comment]
`Focal loss` -> pixel -> targets individual pixels, good for class imbalance. 
`SSIM` ->  patch -> targets the boundaries/edges.
`IoU loss` -> map -> targets the whole foreground object as one unit.

**Comment by JawadTawhidi** [comment]
Ok, thank you so much.

**Comment by JawadTawhidi** [comment]
But one more thing, in the paper they used multi scale structural similarity index(MS-SSIM), but here you used SSIM. It does not change the results? or using any of them is ok?

**Comment by hamidriasat** [comment]
Good catch.
ms-ssim implementation was not available when I implemented the model, so I am not sure how it impacted the results. But now it is. Replace it with [this](https://www.tensorflow.org/api_docs/python/tf/image/ssim_multiscale) and let me know if you feel any changes in the results.

**Comment by JawadTawhidi** [comment]
Yes, I will try it. But when you implemented your repo based on SSIM, did you get the unet 3+ papers results? one more thing is that as I can see in your implementation you did not give weight to losses, it means you treated such that all 3 losses should have equal contribution, but the original paper did not give more details about this if they used weighted losses or just simply used the losses and added their final results.

**Comment by hamidriasat** [comment]
[Here](https://github.com/NVIDIA/DeepLearningExamples/tree/master/TensorFlow2/Segmentation/Contrib/UNet3P#results) from my work, you can see I was able to reproduce results.
Regarding weight as far as I can remember they gave equal weights to all losses.
